### PR TITLE
Refresh Git index prior to fast-forward.

### PR DESCRIPTION
**Pull request by MichaelGrupp**
Fixes https://github.com/vcstools/wstool/issues/77, see the discussion
there for more context.

### CODE REVIEW

File: src/vcstools/git.py [hunk 1]
```diff
@@ -845,12 +845,19 @@ def _do_fast_forward(self, branch_parent, fetch=True, verbose=False):
                 return True
             print("Cannot fast-forward, local repository and remote '%s' have diverged." % branch_parent)
             return False
+        # Refresh the index manually, just as 'git pull' does it.
```
Comment (tkruse) [inline src/vcstools/git.py]: comment here should explain *why*, or link to github issue at least.

**Comment by MichaelGrupp** [comment]
@tkruse Looks like some (unrelated?) "tar" test on Travis failed to download some stuff, maybe you need to restart that build.

**Comment by tkruse** [comment]
Test failure probably caused by 141

**Comment by MichaelGrupp** [comment]
Any update?

**Comment by 130s** [comment]
Like I mentioned in https://github.com/vcstools/vcstools/pull/141#issuecomment-450218309, the issue this PR tries to solve causes my team an issue almost everyday (requires manually workaround). It'd be great this merges in and new release will be made soon.

CI is failing. As suggested at https://github.com/vcstools/vcstools/pull/144#issuecomment-427551430, https://github.com/vcstools/vcstools/pull/141 might have to be processed first (I'm not sure though)?

If any help is needed I'm willing to provide just in general for maintenance BTW (I've been comfortably making releases into ROS for years, although never done ROS-standalone package like vcstools).

**Comment by tfoote** [comment]
reran travis

**Comment by MichaelGrupp** [comment]
Travis is happy

**Comment by 130s** [comment]
Please review  @tkruse / @tfoote / @clalancette.

**Comment by tkruse** [comment]
I can reproduce a similar issue like this (changing file permissions before git reset keep, and changing them back):

    cd /tmp
    git init parent
    cd parent/
    touch foo.txt
    git add foo.txt 
    git commit -m foo
    cd ..
    git clone parent child
    cd parent
    echo 'hello' > foo.txt 
    git add foo.txt 
    git commit -m foo2
    cd ../child/
    chmod u+x foo.txt 
    git fetch
    git reset --keep remotes/origin/master # Error
    chmod u-x foo.txt 
    git reset --keep remotes/origin/master # still Error
    git update-index -q --refresh
    git reset --keep remotes/origin/master # works

Not sure what else can cause the index to go stale like that in the wild (https://stackoverflow.com/questions/1248029/git-pull-error-entry-foo-not-uptodate-cannot-merge suggests filename case on Mac could cause this. https://stackoverflow.com/questions/878554/why-do-i-sometimes-see-an-entry-filename-not-uptodate-cannot-merge-after-a/878742 also mentions this error happens spuriously).

I think the unit tests should be extended with the flow above, even if there is some risk that it cannot be reproduced with all git versions (in that case, tests should be run only for one git version known to reproduce the issue).

Another question to consider here is whether `git update-index -q --refresh` can fail in some situation where we don't want it to fail, or has other undesired side-effects. I think the unit tests protect us mostly from the first, and from the description I believe `--refresh` should not have undesired side-effects. There is also `--really-refresh`, but I believe that's not what we want.

**Comment by MichaelGrupp** [comment]
@tkruse Thanks for looking deeper into this. I'm also not sure if it's a git bug or just something that one has to know/resolve manually when using such low-level plumbing commands. High-level commands like `git merge` refresh the index do this implicitly for a reason (e.g.: https://github.com/git/git/blob/ecbdaf0899161c067986e9d9d564586d4b045d62/builtin/merge.c#L688)

---

One nit: I would use full commands like `git checkout` instead of aliases like `git co` in your test sample.

**Comment by tkruse** [comment]
Feel free to ask for merge in any case, I have no strong objection even without more tests.

I would be happy to accept the git behavior as normal low-level plumbing lacking a refresh step, but since it seems to have an asynchronous one that causes race conditions, I am not sure. We'll see if anyone answers https://marc.info/?l=git&m=154755707120508&w=2

**Comment by MichaelGrupp** [comment]
Feel free to merge, especially since several people already complained.

**Comment by MichaelGrupp** [comment]
Also, this should trigger a new wstool release soon. For anyone that is blocked by the missing index refresh, just calling `wstool status` is a workaround. I did this here for example: https://github.com/googlecartographer/cartographer_ros/commit/8fab3fe4c45fe7d0e30e49cc10779c4feccece89

**Comment by 130s** [comment]
@tfoote @clalancette Please review. @tkruse seems to have approved this at https://github.com/vcstools/vcstools/pull/144#issuecomment-454387684

**Pull request merged**